_source/_posts/2022-04-11-java-cli-device-grant.adoc

```diff
@@ -0,0 +1,412 @@
+
+====
+[source,java]
+----
+long pollUntilMillis = System.currentTimeMillis() + expiresIn.toMillis();
+while (System.currentTimeMillis() < pollUntilMillis) {
+
+    // first sleep, give the user time to log in!
+    System.out.println("Sleeping for " + interval.getSeconds() + " seconds");
+   Thread.sleep(interval.toMillis()); // <.>
+
+    Map<String, String> tokenArgs = Map.of("client_id", clientId,
+            "grant_type","urn:ietf:params:oauth:grant-type:device_code", // <.>
+            "device_code", deviceCode);
+
+    HttpRequest tokenRequest = HttpRequest.newBuilder() // <.>
+            .uri(URI.create(issuer + "/v1/token"))
+            .POST(formBody(tokenArgs))
+            .header("Accept", "application/json")
+            .header("Content-Type", "application/x-www-form-urlencoded")
+            .build();
+
+    HttpResponse<Map> tokenResponse =
+            client.send(tokenRequest, rs -> parseJson(Map.class));
+
+    if (tokenResponse.statusCode() == 200) { // <.>
+            // Do something with the tokens
+            System.out.println("Authorization complete!");
+            System.out.println(tokenResponse.body());
+            break;
+    } else {
+            // error, keep polling until timeout
+            System.out.println("Error: " + tokenResponse.body());
+            System.out.println(tokenResponse);
+        }
+    }
+}
+----
+<.> Wait for the interval to avoid spamming the server, it's going to take a bit for the user to log in anyway.
+<.> Set the `grant_type` to the URN for the device grant.
+<.> Make a request to `/v1/token` to check the authorization status.
+<.> If the response is a `200` inspect the body for an access token, otherwise continue the polling loop.
+====
+
+If the user has not yet logged in, an error is returned with the `authorization_pending` message:
+
+[source,json]
+----
+{
+  "error": "authorization_pending",
+  "error_description": "The device authorization is pending. Please try again later."
+}
+----
+
+Otherwise, if the user has already logged in, a response containing an access token will be returned:
+
+[source,json]
+----
+{
+  "token_type": "Bearer",
+  "expires_in": 3600,
+  "access_token": "...",
+  "scope": "openid profile offline_access",
+  "id_token": "..."
+}
+----
+
+NOTE: This access token would be used to make application-specific REST requests on behalf of the user, typically by setting the `Authorization` header in an HTTP request, for example:
+
+[source,java]
+----
+HttpRequest.newBuilder()
+        .header("Authorization", "Bearer " + accessToken)
+----
+
+== Putting it all together
+
+If you haven't already skipped ahead and ran the finished application, run it now:
+
+[source,shell]
+----
+jbang DeviceGrant.java
+----
+
+This will open a browser with a display a code, make sure this code matches what is displayed in your terminal, and continue logging in!
```

Comment:
```suggestion
This will open a browser with a display code. Make sure this code matches what is displayed in your terminal, and continue logging in.
```
typo